improves binutils search on macos with brew

gitoleg opened the pull request:
the modern brew binutils formula doesn't appear in the user path, and `brew link` refuses to link it even with
`--force` option applied. That's why we need to improve the search and look for `gobjdump` and `c++filt`  utils right in the brew's
cellar folder.

gitoleg committed:
updates conf-binutils

ivg commented on packages/conf-binutils/conf-binutils.0.2/files/find-binutils.ml.in:
```diff
@@ -58,6 +58,14 @@ let split_results fmt =
     | Some s -> split s in
   ksprintf run fmt
 
+let hd_opt xs = List.nth_opt xs 0
```
`List.nth_opt` is available only from 4.05, please refrain from using it.

ivg commented on packages/conf-binutils/conf-binutils.0.2/files/find-binutils.ml.in:
```diff
@@ -122,18 +130,51 @@ let collect_targets objdumps =
+  | "" -> None
+  | x  -> Some x
+
+let rec find_map xs =
```
it is not `find_map` it is just `first_some`, and it is much better to express it as `List.find is_some`

ivg commented on packages/conf-binutils/conf-binutils.0.2/files/find-binutils.ml.in:
```diff
@@ -58,6 +58,14 @@ let split_results fmt =
 
+let hd_opt xs = List.nth_opt xs 0
+
+let take_first fmt =
```
put all commands as thunks in a list and use `List.find is_some` to find the first that returns `Some x`.

ivg commented on packages/conf-binutils/conf-binutils.0.2/files/find-binutils.ml.in:
```diff
@@ -122,18 +130,51 @@ let collect_targets objdumps =
+
+let find_cppfilt () =
+  find_map [of_opam_config "%{cppfilt-path}%";
+            of_env "CPPFILT_PATH";]
```
it should be `CXXFILT_PATH`

ivg commented on packages/conf-binutils/conf-binutils.0.2/files/find-binutils.ml.in:
```diff
@@ -122,18 +130,51 @@ let collect_targets objdumps =
+            of_env "OBJDUMP_PATH";]
+
+let find_cppfilt () =
+  find_map [of_opam_config "%{cppfilt-path}%";
```
`cxxfilt-path`

ivg reviewed (changes requested):
Ideally, try to use the portable subset of OCaml in conf-* files so that we can use it for old OCaml versions. If you need any function from a more modern OCaml or Core, just vendor (the fancy name for copy-paste) it here. But if you find yourself copy-pasting to much code it is probably better just to switch to 4.05 version of OCaml, but don't forget to specify it in the opam file.

Also, there is no need to blindly follow the js style here, you can use exceptions. This is just a script.

gitoleg committed:
refactoring

gitoleg committed:
removes debug messages

ivg reviewed (changes requested):
Your find_some chains are eager, i.e., you first compute all the checks then pick the first that is some. Modify the code such that the check is only run if previous checks have failed. I would suggest turning checks into thunks and modifying your first some to something that looks like,

```ocaml
let rec first_success = function
 | [] -> None
 | f :: fs ->  match f () with
   | None -> first_success fs
   | r -> r
```

gitoleg committed:
makes search less eager

gitoleg committed:
removes unused code

gitoleg committed:
wip

gitoleg committed:
refatored

gitoleg committed:
search in the cellar recursively

thierry-martinez committed:
[new release] clangml.4.2.0

- Support for Clang/LLVM 10.0.1

- Support for all C/C++ Clang attributes (including `_Alignas` (C) and `alignas` (C++) attributes, that have been requested by Damien Rouhling).  See the auto-generated `bootstrap/attributes.ml`.

- class base specifiers now expose a full `qual_type` instead of a mere `ident` string.

- `ident_ref` now carries a `template_arguments` parameter.

- `function_decl` now exposes `inline_specified` and `inlined` fields.

- `function_type` now exposes `ref_qualifier` field.

- `TypeLoc` are used broaderly to compute `qual_type`: in particular, parameters in function types have now correct names.  The type `type_loc` and the functions `Clang.Decl.get_type_loc` and `Clang.Parameter.get_type_loc` are kept for compatibility, but `qual_type` should contain all the informations.

- `Clang.Lazy` exposes a lazy AST, that is to say that each `desc` field is a lazy value that is computed on demand. This is useful to explore large ASTs efficiently (note that Clang parsing itself can still be slow; the lazy part only concerns the conversion into the `Clang.Lazy.Ast` datatypes). Modules `Clang.Ast`, `Clang.Expr`, `Clang.Stmt`, `Clang.Type`, ..., have their lazy counter-parts: `Clang.Lazy.Ast`, `Clang.Lazy.Expr`, `Clang.Lazy.Stmt`, `Clang.Lazy.Type`, etc.

- The common signature exposed by every AST node module (`Clang.Expr`, `Clang.Stmt`, ...) now declares a `hash` function and and a `Hashtbl` module.

- `Clang.format_diagnostics` now prints presumed locations by default and the behavior can be changed through the optional `?options` argument. The implementation now relies on the new `Clang.pp_diagnostic` which is an OCaml reimplementation of `clang_formatDiagnostic` from libclang.

- `Clang.Expr.parse_string` parses strings as C expressions.  (suggested by Damien Rouhling)

- `Clangml_printer` has now moved to `Clang.Printer`. The package `clangml.printer` still exists for compatibility and reexposes the `Clang.Printer` interface.

thierry-martinez committed:
Add support for FreeBSD.

Suggested by Marcello Seri.

mseri committed:
Merge pull request #16827 from thierry-martinez/clangml.4.2.0

[new release] clangml.4.2.0, refl 0.2.1, clangml-transforms 0.26

gitoleg committed:
Merge remote-tracking branch 'upstream/master' into updates-conf-binutils

gitoleg committed:
conf-binutils 0.3

This PR introduces a new version of conf-binutils and constrains all bap
packages to the old ones. The only package beyond bap universe that depends on
`conf-binutils` is `clangml`, and looks like it doesn't depend on our interface, so we leave it untoched.

What's new:

We search for files that have `cxxfilt`, `objdump`, and `readelf` suffixes in the `brew` cellar, if one available, and in the `PATH` constituents. Instead of using tools from `findutils`, `which`, and `mdfind`, as we did before, we search for the matching files manually, using the OCaml `Sys.readdir` function.

The problem is that on macOS with `brew` we can't use which utility for the search: `binutils` formula is marked as `keg-only` and also can shadow system files, so even `brew link --force` doesn't link `binutils` somewhere in the `$PATH`. That's why we have problems with installing bap on macOS: `which gobjdump` doesn't return a path to the file anymore. With this PR this problem will be solved.

Previously, we were creating two `opam` config variables per each tool, e.g. `objdump` and `objdump-targets`. But since all that we do with them on the bap side is concatenating them in one list, there is no need in several variables for one tool. Now we add only one, e.g `objdumps` or `cxxfilts` that contains a list of absolute paths to all discovered files for the given tool. The preferred tool still can be set via `opam` config or environment variable. In this case, it will be first in the paths list.
